ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -175,20 +187,407 @@ let order_f (_, vm_rec) =
+       size, do host selection as below step 2.
+    2. For each anti-affinity VM, select a host which can run it, and which has the least VMs in the
+       same anti-affinity group running on it, for the hosts with the same number of running VMs in
+       that group, pick the one with the least free memory. *)
+let rec vm_anti_affinity_spread_evenly_plan ~__context hosts_increasing
+    vms_increasing vm_can_boot_on_host grp_ranked_hosts mapping =
+  debug "vm_anti_affinity_spread_evenly_plan" ;
+  log_vms_increasing ~__context vms_increasing ;
+  log_hosts_increasing ~__context hosts_increasing ;
+  log_grp_ranked_hosts ~__context grp_ranked_hosts ;
+  match vms_increasing with
+  | [] ->
+      mapping
+  | (vm, vm_size) :: remaining_vms -> (
+      let host_selector hosts =
+        hosts_increasing
+        |> List.filter (fun (h, size) -> List.mem h hosts && size >= vm_size)
+        |> List.map fst
+        |> fst_host_can_run vm_can_boot_on_host vm
+      in
+      match Xapi_vm_helpers.vm_has_anti_affinity ~__context ~vm with
+      | Some (`AntiAffinity group) -> (
+          debug
+            "Spread evenly plan: try to plan for the anti-affinity VM (%s, %s)"
+            (Ref.string_of vm)
+            (Db.VM.get_name_label ~__context ~self:vm) ;
+          try
+            let h =
+              Xapi_vm_helpers.select_host_from_ranked_groups vm host_selector
+                (VMGroupMap.find group grp_ranked_hosts
+                |> IntMap.to_seq
+                |> List.of_seq
+                |> List.fast_sort (fun (a, _) (b, _) -> compare a b)
+                   (* Make sure the ranked list is sorted in terms of vm_count increasingly *)
+                |> List.map snd
+                )
+            in
+            debug "Spread evenly plan: find h: (%s %s)" (Ref.string_of h)
+              (Db.Host.get_name_label ~__context ~self:h) ;
+            let h_size =
+              Int64.sub
+                (List.assoc h hosts_increasing)
+                (List.assoc vm vms_increasing)
+            in
+            debug
+              "Spread evenly plan: choose the host with the least free memory \
+               which can run the vm: %s"
+              (Db.Host.get_name_label ~__context ~self:h) ;
+
+            let hosts_increasing =
+              Binpack.insert bigger_than (h, h_size)
+                (List.remove_assoc h hosts_increasing)
+            in
+            vm_anti_affinity_spread_evenly_plan ~__context hosts_increasing
+              remaining_vms vm_can_boot_on_host
+              (grp_ranked_hosts |> update_anti_affinity_grp_ranked_hosts group h)
+              ((vm, h) :: mapping)
+          with _ ->
+            debug
+              "Spread evenly plan: no host can run the anti-affinity VM, skip \
+               it, continue to plan for the next one." ;
+            vm_anti_affinity_spread_evenly_plan ~__context hosts_increasing
+              remaining_vms vm_can_boot_on_host grp_ranked_hosts mapping
```

Comment:
Since all the available hosts(hosts with enough memory) can't boot the VM, and the final plan needs to be verified with `Xapi_vm_helpers.assert_can_boot_here`, then for this case, we don't need to continue planning anymore.